shotover/src/codec/cassandra.rs

```diff
@@ -221,43 +219,73 @@ impl CassandraDecoder {
 
-                        if envelope_len > payload.len() {
-                            return Err(anyhow!(format!(
-                                "envelope length {} is longer than payload length {}",
-                                envelope_len,
-                                payload.len()
-                            ),));
-                        }
+                    if header_crc24 != computed_crc {
+                        return Err(header_crc_mismatch_error(computed_crc, header_crc24));
+                    }
 
-                        let envelope = payload.split_to(envelope_len);
+                    let compressed_len = (header & 0x1ffff) as usize;
+                    let compressed_payload_end = compressed_len + COMPRESSED_FRAME_HEADER_LENGTH;
 
-                        tracing::debug!(
-                            "{}: incoming cassandra message:\n{}",
-                            self.direction,
-                            pretty_hex::pretty_hex(&envelope)
-                        );
+                    let frame_end = compressed_payload_end + FRAME_TRAILER_LENGTH;
 
-                        envelopes.push(Message::from_bytes(
-                            envelope.freeze(),
-                            crate::message::ProtocolType::Cassandra {
-                                compression: Compression::None,
-                            },
+                    let compressed_payload_crc32 = u32::from_le_bytes(
+                        frame_bytes[compressed_payload_end..frame_end]
+                            .try_into()
+                            .unwrap(),
+                    );
+
+                    let computed_crc =
+                        crc32(&frame_bytes[COMPRESSED_FRAME_HEADER_LENGTH..compressed_payload_end]);
+
+                    if compressed_payload_crc32 != computed_crc {
+                        return Err(payload_crc_mismatch_error(
+                            computed_crc,
+                            compressed_payload_crc32,
                         ));
                     }
 
+                    let self_contained = (header & (1 << 34)) != 0;
+                    if !self_contained {
+                        unimplemented!("Cannot support non-self contained frames yet");
+                    }
+
+                    let uncompressed_length = ((header >> 17) & 0x1ffff) as usize;
+
+                    frame_bytes.advance(COMPRESSED_FRAME_HEADER_LENGTH);
+
+                    let payload = if uncompressed_length == 0 {
+                        // protocol spec 2.2:
+                        // An uncompressed length of 0 signals that the compressed payload should be used as-is
+                        // and not decompressed.
+                        frame_bytes.split_to(compressed_len)
+                    } else {
+                        let mut payload = BytesMut::zeroed(uncompressed_length);
+                        decompress_into(&frame_bytes.split_to(compressed_len), &mut payload)?;
```

Comment:
I figured out my confusion around vec conversions.
While you cant convert a Vec into a BytesMut, you can convert a Vec into a Bytes.
So we can simplify this back into a decompress call:

![image](https://user-images.githubusercontent.com/5120858/233228298-aee73ef1-dec7-4846-be38-467fccdd0777.png)